File: leetcode/145-binary-tree-postorder-traversal.cpp
#include <vector>
#include <stack>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
    public:
        vector<int> postorderTraversal(TreeNode *root) {
            if (root == nullptr) {
                return vector<int>();
            }
            stack<TreeNode*> stk;
            vector<int> result;
            TreeNode *node = root;
            while (node != nullptr) {
                stk.push(node);
                if (node->left != nullptr) {
                    node = node->left;
                } else if (node->right != nullptr) {
                    node = node->right;
                } else {
                    break;
                }
            }
            node = stk.top();
            result.push_back(node->val);
            stk.pop();
            while (!stk.empty()) {
                auto c_node = stk.top();
                if (c_node->right == node || c_node->right == nullptr) {
                    result.push_back(c_node->val);
                    stk.pop();
                    node = c_node;
                } else {
                    c_node = c_node->right;
                    while (c_node != nullptr) {
                        stk.push(c_node);
                        if (c_node->left != nullptr) {
                            c_node = c_node->left;
                        } else if (c_node->right != nullptr) {
                            c_node = c_node->right;
                        } else {
                            break;
                        }
                    }
                    node = stk.top();
                    result.push_back(node->val);
                    stk.pop();
                }
            }
            return result;
        }
};

File: leetcode/145-binary-tree-postorder-traversal.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer[]}
def postorder_traversal(root)
  res = []
  return res if root.nil?
  stk = [[root, 0]]
  while not stk.empty?
    if stk[-1][0].left.nil? and stk[-1][0].right.nil?
      res.push stk.pop[0].val
    elsif stk[-1][1] == 0
      stk[-1][1] = 1
      cnt = stk[-1][0].left
      while not cnt.nil?
        stk.push [cnt, 1]
        cnt = cnt.left
      end
    elsif stk[-1][1] == 1 and not stk[-1][0].right.nil?
      stk[-1][1] = 2
      stk.push [stk[-1][0].right, 0]
    elsif stk[-1][1] == 1
      res.push stk.pop[0].val
    elsif stk[-1][1] == 2
      res.push stk.pop[0].val
    end
  end
  res
end
